### PR TITLE
Feature/update repo status using rx

### PR DESCRIPTION
**Pull request by minidfx**
Hello,

I improved the way for updating the repository status using Reactive Extension keeping the previous rules because on some computers with some repositories the update is very slow.

The work for checking if any changes exists is done in ```TaskPoolScheduler``` and only the work for updating the UI is done on the ```SynchronizationContext``` scheduler.

The application is much reactive because the job isn't made on UI thread like the [Timer](https://msdn.microsoft.com/en-us/library/system.windows.forms.timer(v=vs.110).aspx#Anchor_6).

What do you think?

### CODE REVIEW

**Comment by minidfx** [comment]
I gave the build of the current branch to my colleagues in the company where I work for test purpose.

File: GitUI/UserControls/ToolStripGitStatus.cs [hunk 1]
```diff
@@ -347,28 +298,27 @@ private WorkingStatus CurrentStatus
...
                         _workTreeWatcher.EnableRaisingEvents = true;
                         _gitDirWatcher.EnableRaisingEvents = !_gitDirWatcher.Path.StartsWith(_workTreeWatcher.Path);
-                        ScheduleDeferredUpdate();
                         Visible = true;
-                        return;
-                    default:
-                        throw new NotSupportedException();
+                        break;
+
+                    default: throw new NotImplementedException("Others status aren't implememted yet.");
```
Comment (minidfx) [inline GitUI/UserControls/ToolStripGitStatus.cs]: Typo

File: GitUI/UserControls/ToolStripGitStatus.cs [hunk 2]
```diff
@@ -98,9 +105,41 @@ public ToolStripGitStatus()
...
+                .Where(i => !UICommands.RepoChangedNotifier.IsLocked &&
+                            !GitCommandHelpers.VersionInUse.RaceConditionWhenGitStatusIsUpdatingIndex ||
+                            !Module.IsRunningGitProcess())
+                .Select(i =>
+                {
+                    var command = GitCommandHelpers.GetAllChangedFilesCmd(true, UntrackedFilesMode.Default);
+                    var result = Module.RunGitCmd(command);
+
+                    return GitCommandHelpers.GetAllChangedFilesFromString(Module, result);
+                })
```
Comment (minidfx) [inline GitUI/UserControls/ToolStripGitStatus.cs]: Missing Timeout call for defining the max update time.

**Comment by minidfx** [comment]
I'm not sure whether we still need to catch events of `_workTreeWatcher` and `_gitDirWatcher` because if an interval failed, the next one will success.

I think we can remove `FileSystemWatcher`s in this use case.

**Comment by minidfx** [comment]
I found a problem when git executable is not found and the application is starting.

**Comment by minidfx** [comment]
Some problems occured, the observable is faster than events from filesystemwatcher.

I will spend more time to fix it.

**Comment by EbenZhang** [comment]
After merged your pull request, when doing some operations (check out, commit, reset etc.), it complains `git index.lock already exists` quite frequently. But I'm not 100% sure is it caused by your changes.  and i'm using git 2.7.

**Comment by minidfx** [comment]
I tried with git 2.7 as well and I have more problems with the new one. I started a big refactor in my environment. I will create new pull request later.

**Pull request closed without merging**